obspy/io/reftek/core.py

```diff
@@ -0,0 +1,341 @@
+            st = Stream()
+            for channel_number in np.unique(data['channel_number']):
+                inds = data['channel_number'] == channel_number
+                # channel number of EH/ET packets also equals zero (one of the
+                # three unused bytes in the extended header of EH/ET packets)
+                inds &= data['packet_type'] == b"DT"
+                packets = data[inds]
+
+                # split into contiguous blocks, i.e. find gaps. packet sequence
+                # was sorted already..
+                endtimes = (packets[:-1]["time"] +
+                            packets[:-1]["number_of_samples"] * delta)
+                # check if next starttime matches seamless to last chunk
+                # 1e-3 seconds == 1 millisecond is the smallest time difference
+                # reftek130 format can represent, so anything larger or equal
+                # means a gap/overlap.
+                # for now be conservative and check even more rigorous against
+                # 1e-4 to be on the safe side, but in the gapless data example
+                # the differences are always 0 or -2e-7 (for POSIX timestamps
+                # of order 1e9) which seems like a floating point accuracy
+                # issue for np.float64.
+                gaps = np.abs(packets[1:]["time"] - endtimes) > 1e-4
+                if np.any(gaps):
+                    gap_split_indices = np.nonzero(gaps)[0] + 1
+                    contiguous = np.array_split(packets, gap_split_indices)
+                else:
+                    contiguous = [packets]
+
+                for packets_ in contiguous:
+                    starttime = packets_[0]['time']
+
+                    # Unfortunately the whole data cannot be unpacked with one
+                    # call to libmseed as some payloads do not take the full
+                    # 960 bytes. They are thus padded which would results in
+                    # padded pieces directly in a large array and libmseed
+                    # (understandably) does not support that.
+                    #
+                    # Thus we resort to *tada* pointer arithmetics in Python
+                    # ;-) This is quite a bit faster then correctly casting to
+                    # an integer pointer so its worth it.
+                    #
+                    # Also avoid a data copy.
+                    #
+                    # Writing this directly in C would be about 3 times as fast
+                    # so it might be worth it.
+                    npts = packets_["number_of_samples"].sum()
+                    if headonly:
+                        sample_data = np.array([], dtype=np.int32)
+                    else:
+                        sample_data = np.empty(npts, dtype=np.int32)
+                        pos = 0
+                        s = packets_[0]["payload"][40:].ctypes.data
+                        if len(packets_) > 1:
+                            offset = (
```

Comment:
I added a check that tests that the overall packet array is contiguous, otherwise falling back to the slightly slower but foolproof data unpacking. Data unpacking should be 100% safe now.